plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completion/Completer/FilePath.hs

```diff
@@ -0,0 +1,190 @@
+  resulting PathCompletionInfo would be:
+
+  @
+    partialFileName = "fi"
+    partialFileDir  = "dir1/dir2/fi"
+    ...
+  @
+-}
+data PathCompletionInfo = PathCompletionInfo
+  { partialFileName :: T.Text
+  -- ^ partly written start of next part of path
+  , partialFileDir  :: FilePath
+  -- ^ written part of path
+  , workingDir      :: FilePath
+  -- ^ current working directory of the handled file
+  }
+  deriving (Eq, Show, Read)
+
+{- | Takes a path completion info and returns the list of files
+  in the directory that match the path completion info.
+-}
+listFileCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listFileCompletions recorder complInfo = do
+  let complDir = mkCompletionDirectory complInfo
+  try (evaluate =<< listDirectory complDir) >>= \case
+    Right dirs -> do
+      fixedDirs <-
+        mapM
+          ( \d -> do
+              isDir <- doesDirectoryExist $ mkDirFromCWD complInfo d
+              pure $ if isDir then Posix.addTrailingPathSeparator d else d
+          )
+          dirs
+      pure fixedDirs
+    Left (err :: IOError) -> do
+      logWith recorder Warning $ LogFilePathCompleterIOError complDir err
+      pure []
+
+{- | Returns a list of all (and only) directories in the
+  directory described by path completion info.
+-}
+listDirectoryCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listDirectoryCompletions recorder complInfo = do
+  filepaths <- listFileCompletions recorder complInfo
+  filterM (doesDirectoryExist . mkDirFromCWD complInfo) filepaths
+
+pathCompletionInfoFromCabalPrefixInfo :: CabalPrefixInfo -> PathCompletionInfo
+pathCompletionInfoFromCabalPrefixInfo ctx =
+  PathCompletionInfo
+    { partialFileName = dirNamePrefix
+    , partialFileDir = Posix.addTrailingPathSeparator $ Posix.takeDirectory prefix
+    , workingDir = dir
+    }
+ where
+  prefix = T.unpack $ completionPrefix ctx
+  dirNamePrefix = T.pack $ Posix.takeFileName prefix
+  dir = completionWorkingDir ctx
+
+{- | Returns the directory, the currently handled cabal file is in.
+
+  We let System.FilePath handle the separator syntax since this is used
+  to query filepaths from the system. See Note [Using correct file path separators].
+-}
+mkCompletionDirectory :: PathCompletionInfo -> FilePath
+mkCompletionDirectory complInfo =
+  FP.addTrailingPathSeparator $
+    workingDir complInfo FP.</> (FP.normalise $ partialFileDir complInfo)
+
+{- | Returns the complete filepath for the given partial filepath
+  by combining the current working directory of the cabal file
+  with the given partly written file path.
+
+  Since this is used for completions we use posix separators here.
+  See Note [Using correct file path separators].
+-}
+mkDirFromCWD :: PathCompletionInfo -> FilePath -> FilePath
+mkDirFromCWD complInfo fp =
+  Posix.addTrailingPathSeparator $
+    mkCompletionDirectory complInfo Posix.</> Posix.normalise fp
```

Comment:
Um, you said you need Posix separators but then you call `mkCompletionDirectory` which uses platform-dependent ones?

Reply:
Yes, that is definitely not correct.